lib/lexer/lexer.cpp

```diff
@@ -0,0 +1,353 @@
+	D_ASSERT(handlers[(int)state] != nullptr);
+	return handlers[((int)state)];
+}
+
+Lexer::State Lexer::emptyState() {
+	char ch = chars.peek();
+	if (ch == EOF) {
+		return State::Done;
+	} else if (ch == '\n') {
+		delimit(Token {Token::Type::Newline, Newline()});
+		chars.consume();
+		return State::Empty;
+	} else if (isSpace(ch)) {
+		chars.consume();
+		return State::Empty;
+	} else if (isOperatorStart(ch)) {
+		return State::Operator;
+	} else {
+		return State::Word;
+	}
+}
+
+Lexer::State Lexer::wordState() {
+	State state = State::Word;
+	char ch = chars.peek();
+	if (isSpace(ch) || ch == EOF) {
+		state = State::Empty;
+	} else if (isCommentStart(ch)) {
+		state = State::Comment;
+	} else if (isOperatorStart(ch)) {
+		state = State::Operator;
+	} else if (isCommentStart(ch)) {
+		state = State::Comment;
+	} else if (isBackslash(ch)) {
+		state_data.previous = State::Word;
+		return State::Backslash;
+	} else if (isSingleQuote(ch)) {
+		return State::SingleQuoteStart;
+	} else if (isDoubleQuote(ch)) {
+		return State::DoubleQuoteStart;
+	} else if (ch == '$') {
+		return State::ExpansionStart;
+	} else if (ch == '`') {
+		throw NotImplementedException();
+		// state_data.word.push_back(chars.consume());
+		// return State::CommandSubstitution;
+	} else {
+		state_data.word.push_back(ch);
+		chars.consume();
+		return State::Word;
+	}
+
+	if (!state_data.word.empty()) {
+		delimit(Token {Token::Type::Word, WordToken {std::move(state_data.word)}});
+	}
+	state_data.word.clear();
+	return state;
+}
+
+Lexer::State Lexer::operatorState() {
+	char ch = chars.peek();
+
+	if (isBackslash(ch)) {
+		state_data.previous = State::Operator;
+		return State::Backslash;
+	}
+
+	int matches;
+	if (!isOperatorCharacter(ch)) {
+		matches = 0;
+	} else {
+		matches = Token::prefixOperatorMatches(state_data.word + ch);
+	}
+
+	if (matches >= 1) {
+		// The characters we have now are part of 1 or multiple operators
+		// don't delimit yet, because this does not tell us that it matches an entire operator
+		state_data.word.push_back(ch);
+		chars.consume();
+		return State::Operator;
+	}
+	delimitOperator();
+	return State::Empty;
+}
+
+// Previously it did match a prefix of a valid operator
+// but with the additional character it doesn't;
+// - find out what the previous operator was
+void Lexer::delimitOperator() {
+	auto result = Token::exactOperatorType(state_data.word);
+	D_ASSERT(result.has_value());
+	delimit(Token {result.value(), OperatorToken()});
+	state_data.word.clear();
+}
+
+Lexer::State Lexer::commentState() {
+	char ch = chars.peek();
+	if (isNewline(ch) || ch == EOF) {
+		return State::Empty;
+	}
+	chars.consume();
+	return State::Comment;
+}
+
+Lexer::State Lexer::backslashState() {
+	char ch = chars.consume();
+	D_ASSERT(isBackslash(ch));
+	ch = chars.peek();
+	if (ch == EOF) {
+		// todo: determine behavior
+		throw std::runtime_error("encountered EOF in backslash lexing");
```

Comment:
Is this even possible if we always insert the newline at the end of the input?

Reply:
When reading from a file maybe